pkg/imgpkg/api/describe_bundle.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package api
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	ctlbundle "github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/bundle"
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/lockconfig"
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/registry"
+)
+
+// DescribeOpts Options used when calling the Describe function
+type DescribeOpts struct {
+	Logger      Logger
+	Concurrency int
+}
+
+// Author information from a Bundle
+type Author struct {
+	Name  string `json:"name,omitempty"`
+	Email string `json:"email,omitempty"`
+}
+
+// Website URL where more information of the Bundle can be found
+type Website struct {
+	URL string `json:"url,omitempty"`
+}
+
+// BundleMetadata Extra metadata present in a Bundle
+type BundleMetadata struct {
+	Metadata map[string]string `json:"metadata,omitempty"`
+	Authors  []Author          `json:"authors,omitempty"`
+	Websites []Website         `json:"websites,omitempty"`
+}
+
+// ImageInfo URLs where the image can be found as well as annotations provided in the Images Lock
+type ImageInfo struct {
+	Image       string            `json:"image"`
+	Origin      string            `json:"origin"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// BundleContent Contents present in a Bundle
+type BundleContent struct {
+	Bundles []BundleDescription `json:"bundles,omitempty"`
+	Images  []ImageInfo         `json:"images,omitempty"`
+}
+
+// BundleDescription Metadata and Contents of a Bundle
+type BundleDescription struct {
+	Image       string            `json:"image"`
+	Origin      string            `json:"origin"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Metadata    BundleMetadata    `json:"metadata,omitempty"`
+	Content     BundleContent     `json:"content"`
+}
+
+// DescribeBundle Given a Bundle URL fetch the information about the contents of the Bundle and Nested Bundles
+func DescribeBundle(bundleImage string, opts DescribeOpts, registryOpts registry.Opts) (BundleDescription, error) {
+	reg, err := registry.NewSimpleRegistry(registryOpts)
+	if err != nil {
+		return BundleDescription{}, err
+	}
+	bundle := ctlbundle.NewBundle(bundleImage, reg)
+
+	isBundle, err := bundle.IsBundle()
+	if err != nil {
+		return BundleDescription{}, err
+	}
+
+	if !isBundle {
+		return BundleDescription{}, fmt.Errorf("The image %s is not a bundle", bundleImage)
+	}
+
+	allBundles, _, err := bundle.AllImagesRefs(opts.Concurrency, opts.Logger)
+	if err != nil {
+		return BundleDescription{}, err
+	}
```

Comment:
this to me feels like something that should be happening within the command itself.

Reply:
These checks and validations as well as the creation of the registry is inside this function to ensure that the caller of our API can easily call the function without having to do a big setup before hand